# Mobile App for Lottery Addiction

## Better chance to win: for lottery addicts
This project is about contributing to the development of a mobile app that is meant to help lottery addicts better estimate their chances of winning. Although many people start playing the lottery for fun, for some this activity turns into a habit which eventually escalates into addiction. Like other compulsive gamblers, lottery addicts soon begin spending from their savings and loans, they start to accumulate debts, and eventually engage in desperate behaviors. 

A medical institute that aims to prevent and treat gambling additions by building a dedicated mobile app to help lottery addicts better estimate their chances of winning. We are here to **create the logical core of the app and aclculate probabilities .

For the first version of the app, the focus is on the [6/49 lottery](https://en.wikipedia.org/wiki/Lotto_6/49) and build functions that enable users to answer a few questions.

Since the institute wants us to consider historical data coming from the national 6/49 lottery game in Canada. 

## Core functions

In [45]:
def factorial(n):
    final_product = 1
    for i in range(n, 0, -1):
        final_product *= i
    return final_product

In [46]:
def combinations(n, k):
    return factorial(n) / (factorial(k) * factorial(n-k))

## One ticket probability
Compute the chance of winning the big prize with a single ticket try. Since the 6/49 takes 6 numbers in a row without replacement, we can use the above factorial() and combinations() functions to find the results.

In [47]:
def one_ticket_probability(li):
    n_outcomes = combinations(49, 6)
    prob = 1/n_outcomes * 100
    print("Your chance to win the Big Prize is {:.7f}%. In other word, you will have 1 out of {:,} chances to win.".format(prob, int(n_outcomes)))

In [48]:
test = [2, 43, 22, 23, 11, 5]
one_ticket_probability(test)

Your chance to win the Big Prize is 0.0000072%. In other word, you will have 1 out of 13,983,816 chances to win.


We have theoritically tell users what is the probability of winning the big prize with a single ticket. Next, users should also be albel to compare their ticket agains the historical lottery data in Canada and determine whether they would have ever won by now. 

## Historical data from the Canada 6/49 lottery

The data set can be downloaded from [Kaggle](https://www.kaggle.com/datascienceai/lottery-dataset).

In [49]:
import pandas as pd
ca_649 = pd.read_csv('649.csv')

In [50]:
ca_649.shape

(3665, 11)

In [51]:
ca_649.columns

Index(['PRODUCT', 'DRAW NUMBER', 'SEQUENCE NUMBER', 'DRAW DATE',
       'NUMBER DRAWN 1', 'NUMBER DRAWN 2', 'NUMBER DRAWN 3', 'NUMBER DRAWN 4',
       'NUMBER DRAWN 5', 'NUMBER DRAWN 6', 'BONUS NUMBER'],
      dtype='object')

In [52]:
ca_649.head(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
0      649            1                0  6/12/1982               3   
1      649            2                0  6/19/1982               8   
2      649            3                0  6/26/1982               1   

   NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
0              11              12              14              41   
1              33              36              37              39   
2               6              23              24              27   

   NUMBER DRAWN 6  BONUS NUMBER  
0              43            13  
1              41             9  
2              39            34

In [53]:
ca_649.tail(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3662      649         3589                0  6/13/2018               6   
3663      649         3590                0  6/16/2018               2   
3664      649         3591                0  6/20/2018              14   

      NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
3662              22              24              31              32   
3663              15              21              31              38   
3664              24              31              35              37   

      NUMBER DRAWN 6  BONUS NUMBER  
3662              34            16  
3663              49             8  
3664              48            17

The data set contains 3665 drawings from 1982 to 2018. Each role shows a single drawing.  We can find the six numbers drawn from six separated columns like NUMER DRAWN 1. 

## Functions for historical data check

In [54]:
def extract_numbers(row):
    six_numbers = row[4:10]
    return set(six_numbers.values)

In [55]:
win_nums = ca_649.apply(extract_numbers, axis=1)

In [56]:
def check_historical_occurence(user_nums, win_nums):
    """
    Inputs:
    user_nums: a python list containing the user numbers
    win_nums: a pandas Series containing sets with the winning numbers
    """
    user_nums = set(user_nums)
    check_occurence = user_nums == win_nums
    n_occurrences = check_occurence.sum()
    if n_occurrences == 0:
        print("The combination {} has never occurred. Your chances to win the big prize in the next drawing using the combination {} are 0.0000073%. In other word, you have 1 out of 13,983,816 chances to win".format(user_nums, user_nums))
        
    else:
        print("The combination {} has occurred {} times. Your chances to win the big prize in the nex drawing using the combination {} are 0.0000073%. In other word, you have 1 out of 13,983,816 chances to win.".format(user_nums,n_occurrences ,user_nums))

In [57]:
check_historical_occurence([2, 38, 15, 49, 21, 31], win_nums)

The combination {2, 38, 15, 49, 21, 31} has occurred 1 times. Your chances to win the big prize in the nex drawing using the combination {2, 38, 15, 49, 21, 31} are 0.0000073%. In other word, you have 1 out of 13,983,816 chances to win.


## Multiple ticket probability

As lottery addicts usually play more than one ticket on a single drawing, so our purpose is to help them better estimate their chances of winning. 

Further information from the engineering team:
* The user will input the number of different tickets they want play (without inputting the specific combinations they intend to play).
* Our function will see an integer between 1 and 13,983,816 (the maximum number of different tickets).
* The function should print information about the probability of winning the big prize depending on the number of different tickets played.

In [58]:
def multi_ticket_probability(n_tickets):
    n_combinations = combinations(49, 6)
    probability = n_tickets / n_combinations
    percentage_form = probability * 100
    if n_tickets == 1:
        print('''Your chances to win the big prize with one ticket are {:.6f}%.
In other words, you have a 1 in {:,} chances to win.'''.format(percentage_form, int(n_combinations)))
    
    else:
        combinations_simplified = round(n_combinations / n_tickets)
        print('''Your chances to win the big prize with {:,} different tickets are {:.6f}%.
In other words, you have a 1 in {:,} chances to win.'''.format(n_tickets, percentage_form,
                                                               combinations_simplified))
    

In [63]:
test_inputs = [1, 10, 100, 10000, 1000000, 6991908, 13983816]

for test_input in test_inputs:
    multi_ticket_probability(test_input)
    print('------------------------') # output delimiter

Your chances to win the big prize with one ticket are 0.000007%.
In other words, you have a 1 in 13,983,816 chances to win.
------------------------
Your chances to win the big prize with 10 different tickets are 0.000072%.
In other words, you have a 1 in 1,398,382 chances to win.
------------------------
Your chances to win the big prize with 100 different tickets are 0.000715%.
In other words, you have a 1 in 139,838 chances to win.
------------------------
Your chances to win the big prize with 10,000 different tickets are 0.071511%.
In other words, you have a 1 in 1,398 chances to win.
------------------------
Your chances to win the big prize with 1,000,000 different tickets are 7.151124%.
In other words, you have a 1 in 14 chances to win.
------------------------
Your chances to win the big prize with 6,991,908 different tickets are 50.000000%.
In other words, you have a 1 in 2 chances to win.
------------------------
Your chances to win the big prize with 13,983,816 different ti

So far we have wrote three main functions:
* one_ticket_probability() — calculates the probability of winning the big prize with a single ticket
* check_historical_occurrence() — checks whether a certain combination has occurred in the Canada lottery data set
* multi_ticket_probability() — calculates the probability for any number of tickets between 1 and 13,983,816

Next, we are going to write one more function to allow the users to calculate probabilities for two, three, four or five winning numbers.

In [96]:
def probability_less_6(n_winning_numbers):
    """
    calculates the probability of having two, three, four or five winning numbers
    """
    n_combinations_ticket = combinations(6, n_winning_numbers)
    n_combinations_remaining = combinations(43, 6 - n_winning_numbers)
    successful_outcomes = n_combinations_ticket * n_combinations_remaining
    
    n_combinations_total = combinations(49, 6)    
    probability = successful_outcomes / n_combinations_total
    
    probability_percentage = probability * 100    
    combinations_simplified = round(n_combinations_total/successful_outcomes)    
    print('''Your chances of having {} winning numbers with this ticket are {:.6f}%.
In other words, you have a 1 in {:,} chances to win.'''.format(n_winning_numbers, probability_percentage,
                                                               int(combinations_simplified)))

In [95]:
for test_input in [2, 3, 4, 5]:
    probability_less_6(test_input)
    print('--------------------------') # output delimiter

Your chances of having 2 winning numbers with this ticket are 13.237803%.
In other words, you have a 1 in 8 chances to win.
--------------------------
Your chances of having 3 winning numbers with this ticket are 1.765040%.
In other words, you have a 1 in 57 chances to win.
--------------------------
Your chances of having 4 winning numbers with this ticket are 0.096862%.
In other words, you have a 1 in 1,032 chances to win.
--------------------------
Your chances of having 5 winning numbers with this ticket are 0.001845%.
In other words, you have a 1 in 54,201 chances to win.
--------------------------


## Furture works

For the first version of the app, we have coded four functions
    * one_ticket_probability() — calculates the probability of winning the big prize with a single ticket
    * check_historical_occurrence() — checks whether a certain combination has occurred in the Canada lottery data set
    * multi_ticket_probability() — calculates the probability for any number of of tickets between 1 and 13,983,816
    * probability_less_6() — calculates the probability of having two, three, four or five winning numbers exactly

Possible features for a second version of the app can include:

* Making the outputs even easier to understand by adding fun analogies (for example, we can find probabilities for strange events and compare with the chances of winning in lottery; for instance, we can output something along the lines: "You are 100 times more likely to be the victim of a shark attack than winning the lottery.")

* Combining the one_ticket_probability() and check_historical_occurrence() to output information on probability and historical occurrence at the same time.

* Create a function similar to probability_less_6() which calculates the probability of having at least two, three, four or five winning numbers. Hint: the number of successful outcomes for having at least four winning numbers is the sum of these three numbers:
    - The number of successful outcomes for having four winning numbers exactly
    - The number of successful outcomes for having five winning numbers exactly
    - The number of successful outcomes for having six winning numbers exactly